Add the partial auth feature of Google account to the onboarding setup page

eason9487 opened the pull request:
### Changes proposed in this Pull Request:

Part of #1000 and based on #1060

- Add a new hook `useGoogleAuthorization` for getting Google OAuth URL. It will be reused in the following development.
- Adjust the `<GoogleAccountCard>` and `<AuthorizeGoogleAccountCard>` to add partial auth UI and process.

### Additional note

- [x] Revert https://github.com/woocommerce/google-listings-and-ads/pull/1064/commits/9f465fb3af911db9f2ccf96a4955a994ee7bf329 before merging this PR. See https://github.com/woocommerce/google-listings-and-ads/pull/1064#discussion_r736426466.

### Screenshots:

#### 🎥 . Grant without optional permissions

https://user-images.githubusercontent.com/17420811/138865447-80a2134e-8746-4dc4-9d40-f876e428105e.mp4

#### 🎥 . Grant with "Manage your product listings and accounts for Google Shopping" permission only

https://user-images.githubusercontent.com/17420811/138865473-d077d17d-6335-4760-a04c-93fb53167b9c.mp4

#### 🎥 . Grant with minimal required permissions

https://user-images.githubusercontent.com/17420811/138865491-988e2d5a-2104-4bc7-92a4-653a66b49709.mp4


### Detailed test instructions:

#### Test the re-asking permission flow

1. Disconnect all accounts.
1. Go to the onboarding setup and proceed with the account connection setup.
1. When authorizing plugin with your Google account,
    1. don't tick off any optional permissions
    1. only tick off the "**Manage your product listings and accounts for Google Shopping**" permission.
1. It should display an error tip and re-ask you to provide all permission.
1. Click on the "Allow full access" button, it should bring you to the Google permissions authorization directly.
1. After accepting, it should back to GLA onboarding setup and be able to proceed with the Google Merchant Center account setup.

#### Test the minimal required permission flow
1. Disconnect all accounts.
1. Go to the onboarding setup and proceed with the account connection setup.
1. When authorizing plugin with your Google account, tick off the minimal required "**Manage your product listings and accounts for Google Shopping**" and "**Manage your new site verifications with Google**" permissions
1. Even though the "**Manage your AdWords campaigns**" was not provided, it can still able to proceed with the next setup.

### Changelog entry

> Add - The partial auth feature of Google account to the onboarding setup page.

eason9487 committed:
Add a new hook useGoogleAuthorization for getting Google Oauth URL

eason9487 committed:
Adjust AuthorizeGoogleAccountCard to show partial auth UI

eason9487 committed:
Add a temporary hack for demo

eason9487 commented on js/src/components/google-account-card/authorize-google-account-card.js:
```diff
@@ -21,15 +22,38 @@ const readMoreLink = (
+				u.searchParams.append( 'login_hint', additionalScopeEmail );
+				window.location.href = u.toString();
+				return;
+			}
```
📝  This will be removed after PR #1063 and its dependency PR are merged. And we will have two more overall tests with #1063 when developing the partial auth feature for Google Ads and Google reconnection setups.

ecgan commented on js/src/components/google-account-card/authorize-google-account-card.js:
```diff
@@ -21,15 +22,38 @@ const readMoreLink = (
+	additionalScopeEmail,
+	disabled = false,
+} ) {
+	const isAskingScope = Boolean( additionalScopeEmail );
```
💅  We have this `isAskingScope` that we use to conditionally set things like `description`, `buttonText` and `isDestructive`. For me, the UI (and semantic) is different enough that I would create a separate new account card component (e.g. `RequestFullAccessGoogleAccountCard`), then the code within the components would be more straight forward and simpler, we just need to conditionally show the components at the parent component level.

This is opinionated. We can proceed with what we have here in this PR. 😄

eason9487 replied:
I prefer to keep the current implementation because except for the `isAskingScope` relevant logic, which differs only in text/appearance, the rest parts of component is the same, and they have a relatively larger ratio.

ecgan commented on js/src/components/google-account-card/authorize-google-account-card.scss:
```diff
@@ -0,0 +1,6 @@
+.gla-authorize-google-account-card {
+	&__alter-text {
```
❔ "alter" isn't that specific and meaningful IMO. How about "error"? 🤔 

```suggestion
	&__error-text {
```

ecgan commented on js/src/components/google-account-card/authorize-google-account-card.js:
```diff
@@ -42,7 +66,27 @@ export default function AuthorizeGoogleAccountCard( { disabled = false } ) {
+						),
+						{
+							alert: (
+								<span className="gla-authorize-google-account-card__alter-text" />
```
❔ Would "error" text be better than "alter" text? 

```suggestion
								<span className="gla-authorize-google-account-card__error-text" />
```

ecgan commented on js/src/components/google-account-card/google-account-card.js:
```diff
@@ -8,15 +8,22 @@ import AuthorizeGoogleAccountCard from './authorize-google-account-card';
 	}
 
-	return <AuthorizeGoogleAccountCard disabled={ disabled } />;
+	const additionalScopeEmail = scope.gmcRequired ? undefined : google.email;
```
🚧  When jetpack account is disconnected, there is an error thrown: `TypeError: Cannot read properties of undefined (reading 'email')`.

This seems to fix it:

```suggestion
	const additionalScopeEmail = scope.gmcRequired ? undefined : google?.email;
```

eason9487 replied:
Oops! Thanks for catching this! 👍

ecgan reviewed (changes requested):
Tested as per test instructions. There is only one blocker 🚧  which happens when jetpack account is disconnected. I fixed it locally, and added it as a suggested change in the comment. Other than that, things work as expected.

I left a few other comments which are more opinion-based, not a big issue.

eason9487 committed:
Rename CSS class authorize-google-account-card__alter-text. [skip ci]

Address
- https://github.com/woocommerce/google-listings-and-ads/pull/1064#discussion_r736792064
- https://github.com/woocommerce/google-listings-and-ads/pull/1064#discussion_r736792920

Co-authored-by: Gan Eng Chin <ecgan@users.noreply.github.com>

eason9487 committed:
Rename CSS class authorize-google-account-card__alter-text. [skip ci]

Address https://github.com/woocommerce/google-listings-and-ads/pull/1064#discussion_r736792920

Co-authored-by: Gan Eng Chin <ecgan@users.noreply.github.com>

eason9487 committed:
Fix the accessing undefined bug when JetPack account is not yet connected

Address https://github.com/woocommerce/google-listings-and-ads/pull/1064#discussion_r736803473

Co-authored-by: Gan Eng Chin <ecgan@users.noreply.github.com>

eason9487 committed:
Adjust the condition of additionalScopeEmail to make it more precise

ecgan reviewed (approved):
LGTM 👍

eason9487 committed:
Revert "Add a temporary hack for demo"

This reverts commit 9f465fb3af911db9f2ccf96a4955a994ee7bf329.

eason9487 committed:
Copy the same logic changes of whether disabled GMC card in commit 4ab3e4f7236468fde6f8c6ebe9db45d2ab5a8471

eason9487 commented:
Reverted https://github.com/woocommerce/google-listings-and-ads/pull/1064/commits/9f465fb3af911db9f2ccf96a4955a994ee7bf329, and moved the same logic changes of whether disabled Google Merchant Center account card from the commit 4ab3e4f7236468fde6f8c6ebe9db45d2ab5a8471. Since the logic changes have been reviewed in #1060, will proceed to merge this PR.

![2021-10-28 12 37 10](https://user-images.githubusercontent.com/17420811/139188317-f044cee9-7555-4c93-b7d0-652260a52bca.png)
